clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3130,25 +3145,57 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
       return;
   }
 
-  if (WorkGroupAttr *ExistingAttr = D->getAttr<WorkGroupAttr>()) {
-    ASTContext &Ctx = S.getASTContext();
+  ASTContext &Ctx = S.getASTContext();
+
+  if (!XDimExpr->isValueDependent() && !YDimExpr->isValueDependent() &&
+      !ZDimExpr->isValueDependent()) {
     Optional<llvm::APSInt> XDimVal = XDimExpr->getIntegerConstantExpr(Ctx);
     Optional<llvm::APSInt> YDimVal = YDimExpr->getIntegerConstantExpr(Ctx);
     Optional<llvm::APSInt> ZDimVal = ZDimExpr->getIntegerConstantExpr(Ctx);
-    Optional<llvm::APSInt> ExistingXDimVal = ExistingAttr->getXDimVal(Ctx);
-    Optional<llvm::APSInt> ExistingYDimVal = ExistingAttr->getYDimVal(Ctx);
-    Optional<llvm::APSInt> ExistingZDimVal = ExistingAttr->getZDimVal(Ctx);
 
-    // Compare attribute arguments value and warn for a mismatch.
-    if (ExistingXDimVal != XDimVal || ExistingYDimVal != YDimVal ||
-        ExistingZDimVal != ZDimVal) {
-      S.Diag(AL.getLoc(), diag::warn_duplicate_attribute) << AL;
-      S.Diag(ExistingAttr->getLocation(), diag::note_conflicting_attribute);
+    XDimExpr = checkWorkSizeAttrExpr(S, AL, XDimExpr);
+    YDimExpr = checkWorkSizeAttrExpr(S, AL, YDimExpr);
+    ZDimExpr = checkWorkSizeAttrExpr(S, AL, ZDimExpr);
+
+    if (!XDimExpr || !YDimExpr || !ZDimExpr)
+      return;
+
+    // Skip SEMA if we're in a template, this will be diagnosed later.
+    if (S.getCurLexicalContext()->isDependentContext())
+      return;
+
+    if (AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize) {
```

Comment:
I'm not insisting on changing it as part of this patch, but this is a "code smell" (I suspect we'll change it as part of a larger refactoring effort) -- the function looks like it's generic (it accepts a template parameter) but the functionality isn't actually generic because we have to look to see how the function was instantiated.

Reply:
Done. Removed this